src/libraries/System.Text.Json/Common/ReflectionExtensions.cs

```diff
@@ -3,11 +3,227 @@
+            {
+                case ImmutableDictionaryGenericTypeName:
+                case ImmutableDictionaryGenericInterfaceTypeName:
+                    return ImmutableDictionaryTypeName;
+                case ImmutableSortedDictionaryGenericTypeName:
+                    return ImmutableSortedDictionaryTypeName;
+                default:
+                    // We verified that the type is an immutable collection, so the
+                    // generic definition is one of the above.
+                    return null;
+            }
+        }
+
+        public static string? GetImmutableEnumerableConstructingTypeName(this Type type, bool sourceGenType = false)
+        {
+            Debug.Assert(type.IsImmutableEnumerableType(sourceGenType));
+
+            // Use the generic type definition of the immutable collection to determine
+            // an appropriate constructing type, i.e. a type that we can invoke the
+            // `CreateRange<T>` method on, which returns the desired immutable collection.
+            switch (GetBaseNameFromGenericType(type, sourceGenType))
+            {
+                case ImmutableArrayGenericTypeName:
+                    return ImmutableArrayTypeName;
+                case ImmutableListGenericTypeName:
+                case ImmutableListGenericInterfaceTypeName:
+                    return ImmutableListTypeName;
+                case ImmutableStackGenericTypeName:
+                case ImmutableStackGenericInterfaceTypeName:
+                    return ImmutableStackTypeName;
+                case ImmutableQueueGenericTypeName:
+                case ImmutableQueueGenericInterfaceTypeName:
+                    return ImmutableQueueTypeName;
+                case ImmutableSortedSetGenericTypeName:
+                    return ImmutableSortedSetTypeName;
+                case ImmutableHashSetGenericTypeName:
+                case ImmutableSetGenericInterfaceTypeName:
+                    return ImmutableHashSetTypeName;
+                default:
+                    // We verified that the type is an immutable collection, so the
+                    // generic definition is one of the above.
+                    return null;
+            }
+        }
+
+        private static bool OpenGenericTypesHaveSamePrefix(Type t1, Type t2)
+            => t1.FullName == GetBaseNameFromGenericTypeDef(t2);
+
+        private static string GetBaseNameFromGenericType(Type genericType, bool sourceGenType)
+        {
+            Type genericTypeDef = genericType.GetGenericTypeDefinition();
+            return sourceGenType ? GetBaseNameFromGenericTypeDef(genericTypeDef) : genericTypeDef.FullName!;
+        }
+
+        private static string GetBaseNameFromGenericTypeDef(Type genericTypeDef)
+        {
+            Debug.Assert(genericTypeDef.IsGenericType);
+            string fullName = genericTypeDef.FullName!;
+            int length = fullName.IndexOf("`") + 2;
+            return fullName.Substring(0, length);
```

Comment:
Only works when < 10 generic values? (which is fine for known types, but not things like `Func<,,,,,,,,,,>`)

Reply:
This is related to https://github.com/dotnet/runtime/pull/55566#discussion_r670101418. I think we can remove this manual code altogether; will tackle in follow up.